pkg/ext/stores/useractivity/store.go

```diff
@@ -0,0 +1,258 @@
+package useractivity
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	ext "github.com/rancher/rancher/pkg/apis/ext.cattle.io/v1"
+	v3Legacy "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	v3 "github.com/rancher/rancher/pkg/generated/controllers/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/settings"
+	"github.com/rancher/rancher/pkg/wrangler"
+	extcore "github.com/rancher/steve/pkg/ext"
+	metainternalversion "k8s.io/apimachinery/pkg/apis/meta/internalversion"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/registry/rest"
+)
+
+const (
+	UserActivityNamespace = "cattle-useractivity-data"
+	tokenUserId           = "authn.management.cattle.io/token-userId"
+	SingularName          = "useractivity"
+	PluralName            = "useractivities"
+)
+
+// +k8s:openapi-gen=false
+// +k8s:deepcopy-gen=false
+type Store struct {
+	tokenController v3.TokenController
+	checker         userHandler
+}
+
+var GV = schema.GroupVersion{
+	Group:   "ext.cattle.io",
+	Version: "v1",
+}
+
+var GVK = schema.GroupVersionKind{
+	Group:   GV.Group,
+	Version: GV.Version,
+	Kind:    "UserActivity",
+}
+var GVR = schema.GroupVersionResource{
+	Group:    GV.Group,
+	Version:  GV.Version,
+	Resource: PluralName,
+}
+
+func NewFromWrangler(wranglerCtx *wrangler.Context) *Store {
+	return &Store{
+		tokenController: wranglerCtx.Mgmt.Token(),
+		checker:         &tokenChecker{},
+	}
+}
+
+// GroupVersionKind implements [rest.GroupVersionKindProvider]
+func (t *Store) GroupVersionKind(_ schema.GroupVersion) schema.GroupVersionKind {
+	return GVK
+}
+
+// NamespaceScoped implements [rest.Scoper]
+func (t *Store) NamespaceScoped() bool {
+	return false
+}
+
+// GetSingularName implements [rest.SingularNameProvider]
+func (t *Store) GetSingularName() string {
+	return SingularName
+}
+
+// New implements [rest.Storage]
+func (t *Store) New() runtime.Object {
+	obj := &ext.UserActivity{}
+	obj.GetObjectKind().SetGroupVersionKind(GVK)
+	return obj
+}
+
+// Destroy implements [rest.Storage]
+func (t *Store) Destroy() {
+}
+
+// Create implements [rest.Creator]
+func (uas *Store) Create(ctx context.Context,
+	obj runtime.Object,
+	createValidation rest.ValidateObjectFunc,
+	options *metav1.CreateOptions) (runtime.Object, error) {
+
+	// retrieving useractivity object from raw data
+	objUserActivity, ok := obj.(*ext.UserActivity)
+	if !ok {
+		return nil, fmt.Errorf("error casting runtime object to useractivity")
```

Comment:
This is a general comment regarding all the error returns in the entire store.

AFAIK (from Tom) all the store interface methods have to return kubernetes `apierrors`, not general go errors. While deeper helper functions may return simple go errors what comes out of the interface must not be. See https://github.com/rancher/rancher/pull/47643 for how the token store does this, by properly wrapping the internal errors ...

Useful api errors are `NewBadRequest` (for errors due to user errors in the request), `NewInternalError` (for internal issues not due to user error), `NewNotFound()`, `NewAlreadyExists()`, etc.

Reply:
thanks, I'll refactor the returned errors accordingly asap.